### PR TITLE
[ADD] Reservation, ReservationDao, ReservationDaoImpl

### PR DESCRIPTION
**Pull request by pthutta**

### CODE REVIEW

File: BookingManager-persistence/src/main/java/cz/muni/fi/pa165/brown/entity/Reservation.java [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+    @NotNull
+    @ManyToOne
+    private User user;
+
+    @NotNull
+    @ManyToOne
+    private Room room;
+
+    @NotNull
+    private Date from;
```
Comment (vasekstebra) [inline BookingManager-persistence/src/main/java/cz/muni/fi/pa165/brown/entity/Reservation.java]: Add annotation Temporal(TemporalType.TIMESTAMP)

File: BookingManager-persistence/src/main/java/cz/muni/fi/pa165/brown/entity/Reservation.java [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+
+    @NotNull
+    private Date to;
```
Comment (vasekstebra) [inline BookingManager-persistence/src/main/java/cz/muni/fi/pa165/brown/entity/Reservation.java]: Same annotation as for from attribute

**Review by vasekstebra — CHANGES_REQUESTED** [review]
Please see comments inline. Also please add annotation @Column(nullable=false) for all attributes that have NotNull annotation as per seminar discussion. Good job (y)

**Comment by Dominick1993** [comment]
Just a side note, wouldn't it be better if used something like `java.sql.Timestamp` instead of 'java.util.Date' for the purpose of storing the date?

**Comment by vasekstebra** [comment]
@Dominick1993 In the seminar tasks was used `java.util.Date` with the Temporal annotation to fine tune it to the database datatype. From what I was able to find I didn't find much difference between using either `java.sql.Timestamp` and `java.util.Date` with the `@Temporal(TemporalType.TIMESTAMP)` annotation for a simple project. But according to [this article](http://www.developerscrappad.com/228/java/java-ee/ejb3-jpa-dealing-with-date-time-and-timestamp/) there could be some problems when persisting and retrieving `java.sql.Timestamp` and converting it to `java.util.Date` to use it in the bussines code of the application. So I guess the safest way is to stick with`java.util.Date` and the Temporal annotation.

File: BookingManager-persistence/src/main/java/cz/muni/fi/pa165/brown/entity/Reservation.java (hunk 1 above)
Comment (Dominick1993) [inline BookingManager-persistence/src/main/java/cz/muni/fi/pa165/brown/entity/Reservation.java]: Please change the name of this field to something which does not match SQL keyword exactly (e.g. reservedFrom). While I was implementing the tests I found out that "from" field name caused trouble in SQL statements that hibernate produces, which has caused the `org.hibernate.exception.SQLGrammarException`. Also consider changing `private Date to` field name to match with the newly introduced "from" name.

File: BookingManager-persistence/src/main/java/cz/muni/fi/pa165/brown/entity/Reservation.java [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+
+    @Override
+    public int hashCode() {
+        int result = getId() != null ? getId().hashCode() : 0;
+        result = 31 * result + (getUser() != null ? getUser().hashCode() : 0);
+        result = 31 * result + (getRoom() != null ? getRoom().hashCode() : 0);
+        result = 31 * result + (getFrom() != null ? getFrom().hashCode() : 0);
+        result = 31 * result + (getTo() != null ? getTo().hashCode() : 0);
+        return result;
+    }
```
Comment (Dominick1993) [inline BookingManager-persistence/src/main/java/cz/muni/fi/pa165/brown/entity/Reservation.java:158]: I'd suggest not to compare the reservations by `id`, please adjust the `equals()` and `hashCode()` methods accordingly. I made the same (possible) mistake in my implementation of `Room` entity.

**Comment by pthutta** [comment]
@VaclavStebra Alright, all issues should be fixed now. Thanks for pointing them out.

**Comment by vasekstebra** [comment]
Thanks! It looks good.

**Review by vasekstebra — APPROVED** [review]

**Pull request merged**